feat(grapher): allow authors to specify a preferred missing data strategy

sophiamersmann opened the pull request:
fixes #1177 

Grapher currently implements two different strategies to deal with entities that only have partial data (i.e. entities that are missing data for at least one variable). Here are two charts that exemplify the two strategies that Grapher currently employs:

- [LineChart](https://ourworldindata.org/grapher/production-vs-consumption-co2-emissions?country=~MKD): Consumption-based CO2 emissions are missing for North Macedonia → chart is rendered with missing data (which is usually fine, but in this case problematic as described in #1177)
- [StackedArea](https://ourworldindata.org/grapher/population-breakdown-by-highest-level-of-education-achieved-for-those-aged-15-in?country=~ABW): Aruba does have data for one of the variables, but most variables are missing → No chart is rendered (which is usually a good default to avoid showing misleading charts, but can be undesired in some cases, see comments on Slack [here](https://owid.slack.com/archives/C04PZLMLVFY/p1679351095481309?thread_ts=1678986529.574209&cid=C04PZLMLVFY) and [here](https://owid.slack.com/archives/C04PZLMLVFY/p1679351095481309?thread_ts=1678986529.574209&cid=C04PZLMLVFY))

This PR adds an additional “Missing data” section in the “Data” tab that allows authors to specify a preferred “Missing data strategy”. Three options are available:

- “Automatic” (default): picks a reasonable strategy based on the chart type (this makes the current behaviour explicit)
- “Hide entities with missing data”: Only charts with complete data are rendered. Charts with missing data are not shown – instead, the chart canvas says “No matching data” and it is suggested to change the country/entity
- “Show entities with missing data”: All charts are rendered, including those with missing data.

Caveat: Selecting “Hide charts with missing data” makes #1341 worse as charts for countries/entities with missing data will not be rendered but show up in the select dialog. ~(I made an attempt to fix this issue in this PR.)~

Chart types that hide charts with missing data by default (to do: implement the opposite behaviour):

- [x]  StackedArea
- [x]  StackedBar

Chart types that render charts with missing data by default (to do: implement the opposite behaviour):

- [x]  StackedDiscreteBar
- [x]  DiscreteBar
- [x]  LineChart

Chart types that solve the missing data problem visually (nothing to be done here):

- Marimekko

Doesn’t apply (nothing to be done here):

- ScatterPlot
- TimeScatter
- SlopeChart

Other TODOs:

- [x]  Apply schema changes to schema version 3 once available
- [x]  Hide “Missing Data” section when not applicable?
- [x] Only enable this option for multi-entity charts?

marcelgerber commented on adminSiteClient/EditorCustomizeTab.tsx:
```diff
@@ -376,6 +379,50 @@ class TimelineSection extends React.Component<{ editor: ChartEditor }> {
 }
 
+@observer
+class MissingDataSection extends React.Component<{ editor: ChartEditor }> {
```
I would move all this to `EditorDataTab`, actually.

marcelgerber commented on adminSiteClient/EditorCustomizeTab.tsx:
```diff
@@ -376,6 +379,50 @@ class TimelineSection extends React.Component<{ editor: ChartEditor }> {
+        return (
+            <Section name="Missing data">
+                <SelectField
+                    label="Missing data strategy"
```
May clarify this a bit more:
```suggestion
                    label="Missing data strategy (for when some variables are missing data)"
```

marcelgerber commented on adminSiteClient/EditorCustomizeTab.tsx:
```diff
@@ -376,6 +379,50 @@ class TimelineSection extends React.Component<{ editor: ChartEditor }> {
+    }[] {
+        const missingDataStrategyLabels = {
+            [MissingDataStrategy.auto]: "Automatic",
+            [MissingDataStrategy.hide]: "Hide charts with missing data",
```
We have this slight problem that in our lingo, "chart" usually refers to the whole chart chrome (i.e. grapher as a whole), which is a bit confusing here.
I'm trying to come up with alternative words, but it's tough: plot, entities, visualisation, ...

sophiamersmann replied:
Ahh, okay. I think I'm leaning towards entities then? But it's a difficult one...

sophiamersmann replied:
I said "entities" for now but we can sleep on it :)

marcelgerber commented on packages/@ourworldindata/grapher/src/core/Grapher.tsx:
```diff
@@ -614,6 +616,22 @@ export class Grapher
+
+        // however, sometimes we choose not to render a chart for an entity in case of missing
+        // data. In that case, we want to remove the entity from the selection table as well.
+        if (
+            (this.missingDataStrategy === MissingDataStrategy.hide &&
+                (this.isLineChart ||
+                    this.isDiscreteBar ||
+                    this.isStackedDiscreteBar)) ||
+            (this.missingDataStrategy !== MissingDataStrategy.show &&
+                (this.isStackedArea || this.isStackedBar))
+        ) {
+            return this.inputTable.dropRowsWithErrorValuesForAnyColumn(
+                this.yColumnSlugs
+            )
+        }
```
Could make this into an exported `dropRowsAccordingToMissingDataStrategy(table: OwidTable)` instead, so it can be used by both `tableForSelection` as well as its own `transformTable` method.

sophiamersmann replied:
Yeah, agreed that this is very brittle!

marcelgerber reviewed (commented):
Cool, this works well and is a sound approach 👏🏻

sophiamersmann commented on packages/@ourworldindata/grapher/src/core/Grapher.tsx:
```diff
@@ -614,6 +616,22 @@ export class Grapher
         // entities where data is available)
+
+        // however, sometimes we choose not to render a chart for an entity in case of missing
+        // data. In that case, we want to remove the entity from the selection table as well.
```
Ahh, weird. I'll look into it but I'm also still struggling with MobX, so will likely get back to you for this one...

sophiamersmann commented on adminSiteClient/EditorFeatures.tsx:
```diff
@@ -101,4 +101,15 @@ export class EditorFeatures {
+
+    @computed get canSpecifyMissingDataStrategy() {
+        return (
+            this.grapher.canChangeEntity &&
```
mhh.. yes, I think you're right!

sophiamersmann commented:
Thanks for the thorough review, Marcel, it's appreciated! :)))

sophiamersmann commented:
Lars brought to my attention that this approach sometimes trim the data series. An example is the chart with id 3622.

Default strategy (no hiding)
<img width="797" alt="Screenshot 2023-03-31 at 13 40 58" src="https://user-images.githubusercontent.com/12461810/229111116-b44e5aba-d23d-4482-8699-6c1dc51934e6.png">

Hiding entities when missing data has this undesired side effect:
<img width="797" alt="Screenshot 2023-03-31 at 13 41 08" src="https://user-images.githubusercontent.com/12461810/229111122-2c737767-ee22-4861-b06f-13452272da0c.png">

This happens because we drop every year for which at least one column is missing.

marcelgerber commented:
> Lars brought to my attention that this approach sometimes trims the data series. An example is the chart with id 3622.

Ah, that's a bummer. Not sure how we would want to handle these instead.

sophiamersmann commented:
**Regarding undesired data trimming (see [my comment](https://github.com/owid/owid-grapher/pull/2027#issuecomment-1491802638)):** 
I think it's fine for StackedArea and StackedBar charts to continue using `dropRowsWithErrorValuesForAnyColumn` as we really do want to make sure that we never render a row where values are missing for these chart types (because that could be misleading). That is not the case for LineChart, DiscreteBar and StackedDiscreteBar charts however. Here, showing entities with missing data is usually ok, but sometimes undesired. In these cases, instead of dropping rows individually, we check if one of the columns is missing for an entity and if so, drop all data.

**Regarding the input dialog and `tableForSelection` (see https://github.com/owid/owid-grapher/issues/1341):**
I'm not sure how to address this properly, and the data trimming issue complicated the problem further. I also realised that checking for valid values has to run after interpolation. So, I really can't see a solution that doesn't involve re-running basically all transforms for all available entities... which feels wrong. I think we should decouple this issue from this PR for now, is that's okay with you, Marcel? (If you have an approach in mind, I'd be happy to try to implement that, of course)

marcelgerber commented:
Makes sense.
Anyhow, I would like to wait on this until after Daniel is back, since Daniel has previously thought about missing data a lot.

danyx23 commented:
Great work, @sophiamersmann! I think we should discuss this in the group in the next data viz meeting. Sorry that my input was blocking this for so long :)

sophiamersmann commented:
Sounds good, thanks for checking it out! Looking forward to chatting about this on Wednesday :)

sophiamersmann commented:
Hey @danyx23, I just rebased this onto master, reviewed and tested the code again and updated the PR description – I think this is ready for your review (not urgent of course).

sophiamersmann commented:
Thanks for the review! I wanted to merge this today but it slipped by mind 🤦‍♀️ Will do on Thursday!
